community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java

```diff
@@ -689,6 +696,57 @@ protected long scanForHighId()
         }
     }
 
+    /**
+     * {@link CommonAbstractStore} doesn't use version trailers in the end of the stores after a clean shutdown
+     * anymore. Although {@link RecordFormat} is now pluggable and so a {@link CommonAbstractStore} may be used
+     * to open an older version of the store, one which might have version trailers. This method is used
+     * during opening a store to figure out the highest id by scanning from the end. It's very convenient
+     * if we were aware of the existence of version trailers as to support opening older versions without
+     * problems in this regard. A version trailer may span multiple records in a store which has record size
+     * smaller than the trailer length and so the matching takes that into account in that it can figure
+     * out all possible subsets of the trailer to compare with. Without this method the scan which figures
+     * out highest in use id may mistake version trailer "records" for inUse records, if the inUse bit
+     * happened to be set and so would report too high highest id and reading those higher/trailer records,
+     * trying to interpret them as normal records would fail in random and interesting ways.
+     *
+     * @param cursor {@link PageCursor} to read and compare trailer bytes with.
+     * @param offset offset to start reading the record bytes from the cursor.
+     * @param expectedVersionBytes the whole version trailer as a {@code byte[]}.
+     * @param recordSize record size of records in this store.
+     * @return {@code true} if the record at the offset was just a version trailer "record", otherwise
+     * {@code false} where the id of this record will be set as the highest inUse record in this store.
+     */
+    private boolean isJustLegacyStoreTrailer( PageCursor cursor, int offset, byte[] expectedVersionBytes,
+            int recordSize )
+    {
+        try
+        {
+            for ( int i = 0; i < expectedVersionBytes.length; )
+            {
+                // If the version bytes are bigger than record size then we must also compare with subsets
+                // of those bytes in recordSize chunks
+                boolean mismatch = false;
+                for ( int j = 0; i < expectedVersionBytes.length && j < recordSize; i++, j++ )
+                {
+                    byte b = cursor.getByte( offset + j );
```

Comment:
if expectedVersionBytes.length > recordSize inner for loop start to behave strangely: we will no try to read from second record because of `cursor.getByte( offset + j )` that will start again and `!mismatch`
but i do not know if that is a realistic case :)